metron-platform/metron-writer/src/main/java/org/apache/metron/writer/kafka/KafkaWriter.java

```diff
@@ -156,33 +172,61 @@ public void configure(String sensorName, WriterConfiguration configuration) {
     }
   }
 
+  @Override
+  public void init(Map stormConf, TopologyContext topologyContext, WriterConfiguration config)
+      throws Exception {
+    if(this.zkQuorum != null && this.brokerUrl == null) {
+      try {
+        this.brokerUrl = Joiner.on(",").join(KafkaUtils.INSTANCE.getBrokersFromZookeeper(this.zkQuorum));
+      } catch (Exception e) {
+        throw new IllegalStateException("Cannot read kafka brokers from zookeeper and you didn't specify them, giving up!", e);
+      }
+    }
+    this.kafkaProducer = new KafkaProducer<>(createProducerConfigs());
+  }
+
   public Map<String, Object> createProducerConfigs() {
     Map<String, Object> producerConfig = new HashMap<>();
     producerConfig.put("bootstrap.servers", brokerUrl);
     producerConfig.put("key.serializer", keySerializer);
     producerConfig.put("value.serializer", valueSerializer);
     producerConfig.put("request.required.acks", requiredAcks);
+    producerConfig.put(ProducerConfig.BATCH_SIZE_CONFIG, DEFAULT_BATCH_SIZE);
     producerConfig.putAll(producerConfigs == null?new HashMap<>():producerConfigs);
     producerConfig = KafkaUtils.INSTANCE.normalizeProtocol(producerConfig);
     return producerConfig;
   }
 
   @Override
-  public void init() {
-    if(this.zkQuorum != null && this.brokerUrl == null) {
+  public BulkWriterResponse write(String sensorType, WriterConfiguration configurations,
+      Iterable<Tuple> tuples, List<JSONObject> messages) {
+    BulkWriterResponse writerResponse = new BulkWriterResponse();
+
+    List<Map.Entry<Tuple, Future>> results = new ArrayList<>();
+    int i = 0;
+    for (Tuple tuple : tuples) {
+      JSONObject message = messages.get(i++);
+      Future future = kafkaProducer
+          .send(new ProducerRecord<String, String>(kafkaTopic, message.toJSONString()));
```

Comment:
Should we be more defensive when we transform the message to JSON? Considering the broad use of this class, someone might inject something that causes problems during serialization.

Reply:
It's effectively a HashMap, since that's what the JSONObject class extends. I think the writer should basically write out whatever has been passed to it, meaning that the responsibility would be on the invoking class to sort out whether the HashMap has been constructed correctly in order to properly serialize into a JSON string. I've made a conscious decision in this PR to not make any changes to how we go about managing  serialization and deserialization. Are you worried about security threats, or just garbage data coming from parsers and enrichments, potentially?

Reply:
> I've made a conscious decision in this PR to not make any changes to how we go about managing serialization and deserialization. 

I am not suggesting that we change how we do serialization.  I think we need to wrap the `message.toJSONString()` in a try/catch, so that an exception during serialization is handled as an error and added to the `BulkWriterResponse`, just like the other errors that we handle on lines 218 and 226.  

Right now, an error on one tuple will kill the whole batch.  I would think we would want to handle all errors in the same way.

Reply:
Hey @nickwallen - check out this recent commit and let me know what you think. I added a try/catch for a Throwable on serializing the JSON, add it as an error, and continue the loop. Does that look better? I'm hoping we don't have situations where that toJSON method will throw exceptions, but as you mentioned, if it does, now it won't kill the whole batch.

Reply:
That looks great, thanks.